### ISSUE TITLE
Segmentation fault in builtin:fetchurl builder

### ISSUE THREAD

**Issue by pingiun** [NixOS/nix#3566]
**Describe the bug**

I'm testing out Nix on powerpc64le. I was trying to build a vm image from a basic configuration.nix, but I am getting a segmentation fault when building the bash tar.gz derivation.  

I tried to isolate where things go wrong and captured a coredump of this command:
```
nix build /nix/store/myfzvw3m2fzg7r3xkzr15c0flbc9mxiq-bash-4.4.tar.gz.drv --option build-use-substitutes
```

**Expected behavior**

The bash derivation is build and a binary is generated

**`nix-env --version` output**

`nix-env (Nix) 2.3.4` (this is built from the release tarball from nixos.org)

Using Debian buster 10, with these boost versions:
```
libboost-atomic1.67-dev/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-atomic1.67.0/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-chrono1.67-dev/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-chrono1.67.0/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-context-dev/stable,now 1.67.0.1 ppc64el [installed]
libboost-context1.67-dev/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-context1.67.0/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-date-time1.67-dev/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-date-time1.67.0/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-dev/stable,now 1.67.0.1 ppc64el [installed]
libboost-serialization1.67-dev/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-serialization1.67.0/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-system-dev/stable,now 1.67.0.1 ppc64el [installed]
libboost-system1.67-dev/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-system1.67.0/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-thread-dev/stable,now 1.67.0.1 ppc64el [installed]
libboost-thread1.67-dev/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost-thread1.67.0/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
libboost1.67-dev/stable,now 1.67.0-13+deb10u1 ppc64el [installed,automatic]
```

**Additional context**

A backtrace is attached below, I can also make the nix binary and coredump available if necessary.

<details><summary>Backtrace</summary>
<pre>
(gdb) bt full
#0  0x30090510060c0514 in ?? ()
No symbol table info available.
#1  0x00007fffb41d7a74 in boost::context::detail::fiber_unwind (t=...) at /usr/include/boost/context/fiber_fcontext.hpp:58
No locals.
#2  0x00007fffb41d7578 in boost::context::fiber::resume() && (this=<optimized out>) at /usr/include/c++/8/bits/move.h:149
        __PRETTY_FUNCTION__ = "boost::context::fiber boost::context::fiber::resume() &&"
#3  boost::coroutines2::detail::pull_coroutine<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::control_block::<lambda(boost::context::fiber&&)>::operator() (c=..., this=<optimized out>) at /usr/include/boost/coroutine2/detail/pull_control_block_cc.ipp:95
        synthesized_cb = {c = {fctx_ = 0x0}, other = 0x7fffb096ffc0, state = boost::coroutines2::detail::state_t::none, except = {_M_exception_object = 0x0}}
        synthesized = {cb_ = 0x7fffb096fdd8}
        fn_ = <optimized out>
        this = <optimized out>
        fn_ = <optimized out>
        this = <optimized out>
        synthesized_cb = <optimized out>
        synthesized = <optimized out>
        fn = <optimized out>
#4  std::__invoke_impl<boost::context::fiber, boost::coroutines2::detail::pull_coroutine<T>::control_block::control_block(boost::context::preallocated, StackAllocator&&, Fn&&) [with StackAllocator = boost::context::basic_fixedsize_stack<boost::context::stack_traits>; Fn = nix::sinkToSource(std::function<void(nix::Sink&)>, std::function<void()>)::SinkToSource::read(unsigned char*, size_t)::<lambda(boost::coroutines2::coroutine<std::__cxx11::basic_string<char> >::push_type&)>; T = std::__cxx11::basic_string<char>]::<lambda(boost::context::fiber&&)>&, boost::context::fiber> (__f=...) at /usr/include/c++/8/bits/invoke.h:60
No locals.
#5  std::__invoke<boost::coroutines2::detail::pull_coroutine<T>::control_block::control_block(boost::context::preallocated, StackAllocator&&, Fn&&) [with StackAllocator = boost::context::basic_fixedsize_stack<boost::context::stack_traits>; Fn = nix::sinkToSource(std::function<void(nix::Sink&)>, std::function<void()>)::SinkToSource::read(unsigned char*, size_t)::<lambda(boost::coroutines2::coroutine<std::__cxx11::basic_string<char> >::push_type&)>; T = std::__cxx11::basic_string<char>]::<lambda(boost::context::fiber&&)>&, boost::context::fiber> (__fn=...)
    at /usr/include/c++/8/bits/invoke.h:96
No locals.
#6  std::invoke<boost::coroutines2::detail::pull_coroutine<T>::control_block::control_block(boost::context::preallocated, StackAllocator&&, Fn&&) [with StackAllocator = boost::context::basic_fixedsize_stack<boost::context::stack_traits>; Fn = nix::sinkToSource(std::function<void(nix::Sink&)>, std::function<void()>)::SinkToSource::read(unsigned char*, size_t)::<lambda(boost::coroutines2::coroutine<std::__cxx11::basic_string<char> >::push_type&)>; T = std::__cxx11::basic_string<char>]::<lambda(boost::context::fiber&&)>&, boost::context::fiber> (__fn=...)
    at /usr/include/c++/8/functional:82
No locals.
#7  boost::context::detail::fiber_record<boost::context::fiber, boost::context::basic_fixedsize_stack<boost::context::stack_traits>, boost::coroutines2::detail::pull_coroutine<T>::control_block::control_block(boost::context::preallocated, StackAllocator&&, Fn&&) [with StackAllocator = boost::context::basic_fixedsize_stack<boost::context::stack_traits>; Fn = nix::sinkToSource(std::function<void(nix::Sink&)>, std::function<void()>)::SinkToSource::read(unsigned char*, size_t)::<lambda(boost::coroutines2::coroutine<std::__cxx11::basic_string<char> >::push_type&)>; T = std::__cxx11::basic_string<char>]::<lambda(boost::context::fiber&&)> >::run (fctx=<optimized out>, this=<optimized out>) at /usr/include/boost/context/fiber_fcontext.hpp:141
        c = {fctx_ = 0x7fffb096fe28}
        c = <optimized out>
#8  boost::context::detail::fiber_entry<boost::context::detail::fiber_record<boost::context::fiber, boost::context::basic_fixedsize_stack<boost::context::stack_traits>, boost::coroutines2::detail::pull_coroutine<T>::control_block::control_block(boost::context::preallocated, StackAllocator&&, Fn&&) [with StackAllocator = boost::context::basic_fixedsize_stack<boost::context::stack_traits>; Fn = nix::sinkToSource(std::function<void(nix::Sink&)>, std::function<void()>)::SinkToSource::read(unsigned char*, size_t)::<lambda(boost::coroutines2::coroutine<std::__cxx11::basic_string<char> >::push_type&)>; T = std::__cxx11::basic_string<char>]::<lambda(boost::context::fiber&&)> > >(boost::context::detail::transfer_t) (t=...)
    at /usr/include/boost/context/fiber_fcontext.hpp:80
        __PRETTY_FUNCTION__ = "void boost::context::detail::fiber_entry(boost::context::detail::transfer_t) [with Rec = boost::context::detail::fiber_record<boost::context::fiber, boost::context::basic_fixedsize_stack<boost::contex"...
        rec = <optimized out>
#9  0x00007fffb474186c in make_fcontext () from /lib/powerpc64le-linux-gnu/libboost_context.so.1.67.0
No symbol table info available.
</pre>
</details>

**Comment by edolstra** [comment]
#2517 might be related.

**Comment by r-burns** [comment]
I believe this was caused by a bug in boost context, which is resolved as of boost 1.71. Can you try again with boost 1.71 or above and see if that fixes this issue? Alternatively you can cherry-pick the fix to older versions, as it applies cleanly: https://github.com/boostorg/context/commit/d4608a4e8ba2e8a0bdd45613243e6c53a23d23dc

**Comment by pingiun** [comment]
I cannot try this any more because I do no longer have the powerpc machine available  :/
